3. Составить программу решения задачи коммивояжера для графа,
заданного матрицей смежности:

In [ ]:
import math as m
import numpy as np

In [1]:
import math

def tsp_branch_and_bound(graph):
    n = len(graph)
    visited = [False] * n
    best_cost = math.inf
    best_path = []

    def bound(current_cost, level):
        min_edges = []
        for row_idx in range(n):
            if not visited[row_idx]:
                min_edge = min((graph[row_idx][i] for i in range(n) if not visited[i] and graph[row_idx][i] != math.inf), default=0)
                min_edges.append(min_edge)
            else:
                min_edges.append(0)
        return current_cost + sum(min_edges[:n - level])

    def solve(current_node, level, current_path, current_cost):
        nonlocal best_cost, best_path

        if level == n:
            total_cost = current_cost + graph[current_node][0]
            if total_cost < best_cost:
                best_cost = total_cost
                best_path = current_path + [0]
            return

        for next_node in range(n):
            if not visited[next_node] and graph[current_node][next_node] != math.inf:
                next_cost = current_cost + graph[current_node][next_node]
                if bound(next_cost, level) < best_cost:
                    visited[next_node] = True
                    solve(next_node, level + 1, current_path + [next_node], next_cost)
                    visited[next_node] = False

    visited[0] = True
    solve(0, 1, [0], 0)

    return best_cost, best_path

distance_matrix = [
    [math.inf, 7, 12, 25, 10],
    [10, math.inf, 9, 5, 11],
    [13, 8, math.inf, 6, 4],
    [6, 11, 15, math.inf, 15],
    [5, 9, 12, 17, math.inf]
]

best_cost, best_path = tsp_branch_and_bound(distance_matrix)
print("Минимальная стоимость:", best_cost)
print("Оптимальный путь:", best_path)

Минимальная стоимость: 36
Оптимальный путь: [0, 1, 3, 2, 4, 0]


4. Решить задачу коммивояжера для задач, представленных в следующем разделе (варианты заданий).

Пусть граф задан матрицей смежности размером 5×5, в
которой помимо диагональных элементов удалены еще
два ребра


In [ ]:
distance_matrix = [
    [m.inf, m.inf, m.inf, 25, 10],
    [10, m.inf, 9, 5, 11],
    [13, 8, m.inf, 6, 4],
    [6, 11, 15, m.inf, 15],
    [5, 9, 12, 17, m.inf]
]

best_cost, best_path = tsp_branch_and_bound(distance_matrix)
print("Минимальная стоимость:", best_cost)
print("Оптимальный путь:", best_path)


Минимальная стоимость: 40
Оптимальный путь: [0, 4, 1, 2, 3, 0]


In [2]:
tape = ['λ', 'λ', 'λ', '0', '1', '1', '0', '1', 'λ', 'λ', 'λ']

def turing_machine(tape):
    head = 0
    state = 'start'
    binary_string = ''

    while state != 'halt':
        if state == 'start':
            if tape[head] == 'λ':
                head += 1
            else:
                state = 'read'
        elif state == 'read':
            if tape[head] in ['0', '1']:
                binary_string += tape[head]
                head += 1
            else:
                state = 'process'
        elif state == 'process':
            complement = ''.join('1' if bit == '0' else '0' for bit in binary_string)
            result = ['λ', 'λ', 'λ'] + list(complement) + ['*'] + list(complement) + ['λ', 'λ', 'λ']
            state = 'halt'

    return result

print(''.join(turing_machine(tape)))

λλλ10010*10010λλλ


In [3]:
tape = ['λ', 'λ', 'λ', '0', '1', '1', '0', '1', 'λ', 'λ', 'λ']

def turing_machine(tape):
    binary_string = ''
    head = 0

    while head < len(tape) and tape[head] == 'λ':
        head += 1

    while head < len(tape) and tape[head] in ['0', '1']:
        binary_string += tape[head]
        head += 1

    complement = ''.join('1' if bit == '0' else '0' for bit in binary_string)
    result = ['λ', 'λ', 'λ'] + list(complement) + ['*'] + list(complement) + ['λ', 'λ', 'λ']

    return result

print(''.join(turing_machine(tape)))

λλλ10010*10010λλλ


# Отчет по лабораторной работе

## 1. Решение задачи коммивояжера (TSP) методом 'ветвей и границ'

**Описание задачи:**
Задача коммивояжера (TSP) заключается в нахождении кратчайшего возможного маршрута, который проходит через каждый город ровно один раз и возвращается в начальный город. В данном случае, граф задан матрицей смежности.

**Реализация:**
Использован алгоритм 'ветвей и границ' для поиска оптимального маршрута. Функция `tsp_branch_and_bound` принимает матрицу смежности в качестве входных данных и возвращает минимальную стоимость маршрута и оптимальный путь.

**Исходные данные (матрица смежности):**
```python
distance_matrix = [
    [math.inf, 7, 12, 25, 10],
    [10, math.inf, 9, 5, 11],
    [13, 8, math.inf, 6, 4],
    [6, 11, 15, math.inf, 15],
    [5, 9, 12, 17, math.inf]
]
```

**Результаты выполнения (первый случай):**
- Минимальная стоимость: 36
- Оптимальный путь: `[0, 1, 3, 2, 4, 0]`

**Исходные данные (второй случай, с измененной матрицей):**
```python
distance_matrix = [
    [m.inf, m.inf, m.inf, 25, 10],
    [10, m.inf, 9, 5, 11],
    [13, 8, m.inf, 6, 4],
    [6, 11, 15, m.inf, 15],
    [5, 9, 12, 17, m.inf]
]
```

**Результаты выполнения (второй случай):**
- Минимальная стоимость: 40
- Оптимальный путь: `[0, 4, 1, 2, 3, 0]`

## 2. Моделирование машины Тьюринга

**Описание задачи:**
Моделирование машины Тьюринга, которая принимает на вход двоичную строку, а затем вычисляет ее побитовое дополнение и записывает его дважды на ленту, разделяя символом `*`.

**Реализация:**
Две версии функции `turing_machine` были предоставлены. Обе функции считывают двоичную строку с ленты, вычисляют ее побитовое дополнение (инвертируют '0' и '1') и формируют новую ленту с результатом. Символ 'λ' используется для обозначения пустых ячеек ленты.

**Исходные данные (лента):**
```python
tape = ['λ', 'λ', 'λ', '0', '1', '1', '0', '1', 'λ', 'λ', 'λ']
```

**Результаты выполнения (обе версии):**
- Выходная лента: `λλλ10010*10010λλλ`

**Вывод:**
Обе задачи были успешно реализованы и продемонстрировали ожидаемое поведение. Алгоритм 'ветвей и границ' эффективно находит оптимальные маршруты для TSP, а симуляция машины Тьюринга корректно выполняет операцию побитового дополнения и форматирования ленты.